torch/_higher_order_ops/associative_scan.py

```diff
@@ -0,0 +1,173 @@
+from typing import Callable
+
+import torch
+
+import torch._prims_common as utils
+import torch._subclasses.functional_tensor
+
+import torch.utils._pytree as pytree
+
+from torch._C import DispatchKey
+from torch._C._functorch import _add_batch_dim, get_unwrapped, maybe_get_bdim
+from torch._higher_order_ops.utils import (
+    _maybe_run_with_interpreter,
+    _set_compilation_env,
+    autograd_not_implemented,
+    unique_graph_id,
+)
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.fx.experimental.proxy_tensor import (
+    disable_proxy_modes_tracing,
+    make_fx,
+    ProxyTorchDispatchMode,
+    track_tensor_tree,
+)
+
+aten = torch._ops.ops.aten
+
+
+def associative_scan(
+    input: torch.Tensor,
+    dim: int,
+    combine_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+) -> torch.Tensor:
+    r"""
+    Performs an inclusive scan with an associative pointwise combine function.
+
+    .. warning::
+        `torch.associative_scan` is a prototype feature in PyTorch. It currently
+        does not support autograd and you may run into miscompiles.
+        Read more about feature classification at:
+        https://pytorch.org/blog/pytorch-feature-classification-changes/#prototype
+
+    This operator requires runtime code generation and so requires support for
+    ``torch.compile``. Further, only CUDA device codegen is supported at the moment.
```

Comment:
Our other control flow ops execute some python version of the op when torch.compile isn't available. Is there a reason to diverge?

Reply:
An inefficient eager implementation is possible, but hasn't been written yet.